From eea280b1f1d9d66c903014fd6703f401eb119a87 Mon Sep 17 00:00:00 2001
Message-Id: <eea280b1f1d9d66c903014fd6703f401eb119a87.1504913100.git.jan.steffens@gmail.com>
From: "Jan Alexander Steffens (heftig)" <jan.steffens@gmail.com>
Date: Sat, 9 Sep 2017 00:22:13 +0200
Subject: [PATCH] Use OVER operator for drawing text

For opaque text, SOURCE is not any different from OVER. However, when
drawing color glyphs (which consist of RGBA pixels instead of strokes)
SOURCE's handling of alpha is not what we want.

I stumbled across this because cairo 1.15.8 seems to clear the surface
before drawing color emoji if the operator is SOURCE, deleting every-
thing drawn before. Arguably, the area outside the glyph bounds should
not be touched, but even if this is a cairo bug the problem of alpha
within the glyph remains.
---
 libi3/font.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libi3/font.c b/libi3/font.c
index 81091ea74a31da11..aef8427b5396c3d8 100644
--- a/libi3/font.c
+++ b/libi3/font.c
@@ -105,7 +105,7 @@ static void draw_text_pango(const char *text, size_t text_len,
         pango_layout_set_text(layout, text, text_len);
 
     /* Do the drawing */
-    cairo_set_operator(cr, CAIRO_OPERATOR_SOURCE);
+    cairo_set_operator(cr, CAIRO_OPERATOR_OVER);
     cairo_set_source_rgb(cr, pango_font_red, pango_font_green, pango_font_blue);
     pango_cairo_update_layout(cr, layout);
     pango_layout_get_pixel_size(layout, NULL, &height);
-- 
2.14.1

